File: tests/test-rng.py
#!/usr/bin/env python3
"""Parity of the PyTorch CPU generator the acoustic noise comes from.

Owns both sides: draws the reference stream on a seeded CPU generator, runs
the GGML harness on the same seed and count, compares. The uniform stream is
exact, the normal one carries the Box-Muller rounding.
Run from the tests/ directory.

Usage:
    ./test-rng.py
"""
import os
import subprocess
import sys

import numpy as np
import torch

BIN = "../build/test-rng"
TMP = "tmp"

# kind, seed, count, tolerance
CASES = [("uniform", 42, 4096, 0.0), ("normal", 1234, 12864, 1e-5)]


def report(label, ref, got, max_rel):
    if ref.size != got.size:
        print("[Parity] %s FAIL size %d vs %d" % (label, ref.size, got.size))
        return False
    rel = float(np.sqrt(((ref - got) ** 2).mean()) / (np.sqrt((ref ** 2).mean()) + 1e-20))
    mx = float(np.abs(ref - got).max())
    ok = rel <= max_rel
    print("[Parity] %s: rel rms %.3e max abs %.3e %s" % (label, rel, mx, "OK" if ok else "FAIL"))
    return ok


def main():
    os.chdir(os.path.dirname(os.path.abspath(__file__)))
    os.makedirs(TMP, exist_ok=True)

    ok = True
    for kind, seed, count, tol in CASES:
        generator = torch.Generator(device="cpu").manual_seed(seed)
        draw = torch.rand if kind == "uniform" else torch.randn
        ref = draw((count,), dtype=torch.float32, device="cpu", generator=generator).numpy().astype("float32")

        subprocess.run([BIN, str(seed), str(count), kind, TMP + "/rng.bin"], check=True)

        got = np.fromfile(TMP + "/rng.bin", dtype="float32")
        ok = report("rng-" + kind, ref, got, tol) and ok

    sys.exit(0 if ok else 1)


main()

File: tests/test-rng.cpp
// test-rng.cpp: PyTorch CPU generator parity harness
//
// Draws the uniform stream or the standard normal fill of a seeded CPU
// generator and writes them as raw f32, for comparison against torch.

#include "torch-cpu-rng.h"

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <vector>

int main(int argc, char ** argv) {
    if (argc != 5) {
        fprintf(stderr, "usage: %s seed count uniform|normal out.bin\n", argv[0]);
        return 1;
    }

    uint64_t     seed     = strtoull(argv[1], nullptr, 10);
    int64_t      count    = atoll(argv[2]);
    const char * kind     = argv[3];
    const char * out_path = argv[4];

    bool normal = strcmp(kind, "normal") == 0;
    if (!normal && strcmp(kind, "uniform") != 0) {
        fprintf(stderr, "[Test-RNG] kind must be uniform or normal\n");
        return 1;
    }
    if (count < 16) {
        fprintf(stderr, "[Test-RNG] count must be at least 16\n");
        return 1;
    }

    std::vector<float> values((size_t) count);
    if (normal) {
        torch_cpu_randn(seed, values.data(), count);
    } else {
        std::mt19937 engine = torch_cpu_generator(seed);
        for (int64_t i = 0; i < count; i++) {
            values[(size_t) i] = torch_cpu_uniform(engine);
        }
    }

    FILE * out = fopen(out_path, "wb");
    if (!out || fwrite(values.data(), sizeof(float), values.size(), out) != values.size()) {
        fprintf(stderr, "[Test-RNG] cannot write %s\n", out_path);
        return 1;
    }
    fclose(out);

    fprintf(stderr, "[Test-RNG] seed %llu, %lld %s values\n", (unsigned long long) seed, (long long) count, kind);
    return 0;
}
